{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "asian-valentine",
   "metadata": {},
   "source": [
    "# Gene expression dataset (Golub et al.)\n",
    "## Molecular Classification of Cancer by Gene Expression Monitoring"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "vanilla-warner",
   "metadata": {},
   "source": [
    "### Context\n",
    "\n",
    "This dataset comes from a proof-of-concept study published in 1999 by Golub et al. It showed how new cases of cancer could be classified by gene expression monitoring (via DNA microarray) and thereby provided a general approach for identifying new cancer classes and assigning tumors to known classes. These data were used to classify patients with acute myeloid leukemia (AML) and acute lymphoblastic leukemia (ALL).\n",
    "\n",
    "### Content\n",
    "\n",
    "Golub et al \"Molecular Classification of Cancer: Class Discovery and Class\n",
    "Prediction by Gene Expression Monitoring\"\n",
    "\n",
    "There are two datasets containing the initial (training, 38 samples) and independent (test, 34 samples) datasets used in the paper. These datasets contain measurements corresponding to ALL and AML samples from Bone Marrow and Peripheral Blood. Intensity values have been re-scaled such that overall intensities for each chip are equivalent. (Source: https://www.kaggle.com/crawford/gene-expression)\n",
    "\n",
    "### Acknowledgements\n",
    "\n",
    "Molecular Classification of Cancer: Class Discovery and Class Prediction by Gene Expression\n",
    "\n",
    "Science 286:531-537. (1999). Published: 1999.10.14\n",
    "\n",
    "T.R. Golub, D.K. Slonim, P. Tamayo, C. Huard, M. Gaasenbeek, J.P. Mesirov, H. Coller, M. Loh, J.R. Downing, M.A. Caligiuri, C.D. Bloomfield, and E.S. Lander\n",
    "\n",
    "These datasets have been converted to a comma separated value files (CSV).\n",
    "\n",
    "### Inspiration\n",
    "\n",
    "These datasets are great for classification problems. The original authors used the data to classify the type of cancer in each patient by their gene expressions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prompt-aaron",
   "metadata": {},
   "source": [
    "## My approach to analysing this dataset @@@@@@@\n",
    "After reading [Golub et al](https://science.sciencemag.org/content/286/5439/531)'s paper, I will try to reproduce their findings using different classification algorithms. The authors found 50 informative genes as predictors of ALL/AML cancer type.\n",
    "\n",
    "@@@@@@@@@@@ explain my goal here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "artificial-drawing",
   "metadata": {},
   "source": [
    "## Exploratory Data Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "nutritional-invention",
   "metadata": {},
   "source": [
    "take a look:\n",
    "https://www.kaggle.com/nageshsingh/classification-of-cancer # drops 'call' columns!\n",
    "https://www.kaggle.com/varimp/gene-expression-classification # drops 'call' columns!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "freelance-porcelain",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "extra-lebanon",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = pd.read_csv('data_set_ALL_AML_train.csv')\n",
    "test_data = pd.read_csv('data_set_ALL_AML_independent.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "useful-safety",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Description</th>\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>1</th>\n",
       "      <th>call</th>\n",
       "      <th>2</th>\n",
       "      <th>call.1</th>\n",
       "      <th>3</th>\n",
       "      <th>call.2</th>\n",
       "      <th>4</th>\n",
       "      <th>call.3</th>\n",
       "      <th>...</th>\n",
       "      <th>29</th>\n",
       "      <th>call.33</th>\n",
       "      <th>30</th>\n",
       "      <th>call.34</th>\n",
       "      <th>31</th>\n",
       "      <th>call.35</th>\n",
       "      <th>32</th>\n",
       "      <th>call.36</th>\n",
       "      <th>33</th>\n",
       "      <th>call.37</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFFX-BioB-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-5_at</td>\n",
       "      <td>-214</td>\n",
       "      <td>A</td>\n",
       "      <td>-139</td>\n",
       "      <td>A</td>\n",
       "      <td>-76</td>\n",
       "      <td>A</td>\n",
       "      <td>-135</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>A</td>\n",
       "      <td>-318</td>\n",
       "      <td>A</td>\n",
       "      <td>-32</td>\n",
       "      <td>A</td>\n",
       "      <td>-124</td>\n",
       "      <td>A</td>\n",
       "      <td>-135</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFFX-BioB-M_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-M_at</td>\n",
       "      <td>-153</td>\n",
       "      <td>A</td>\n",
       "      <td>-73</td>\n",
       "      <td>A</td>\n",
       "      <td>-49</td>\n",
       "      <td>A</td>\n",
       "      <td>-114</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-114</td>\n",
       "      <td>A</td>\n",
       "      <td>-192</td>\n",
       "      <td>A</td>\n",
       "      <td>-49</td>\n",
       "      <td>A</td>\n",
       "      <td>-79</td>\n",
       "      <td>A</td>\n",
       "      <td>-186</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFFX-BioB-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-3_at</td>\n",
       "      <td>-58</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>A</td>\n",
       "      <td>-307</td>\n",
       "      <td>A</td>\n",
       "      <td>265</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>-95</td>\n",
       "      <td>A</td>\n",
       "      <td>49</td>\n",
       "      <td>A</td>\n",
       "      <td>-37</td>\n",
       "      <td>A</td>\n",
       "      <td>-70</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFFX-BioC-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-5_at</td>\n",
       "      <td>88</td>\n",
       "      <td>A</td>\n",
       "      <td>283</td>\n",
       "      <td>A</td>\n",
       "      <td>309</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>193</td>\n",
       "      <td>A</td>\n",
       "      <td>312</td>\n",
       "      <td>A</td>\n",
       "      <td>230</td>\n",
       "      <td>P</td>\n",
       "      <td>330</td>\n",
       "      <td>A</td>\n",
       "      <td>337</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFFX-BioC-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-3_at</td>\n",
       "      <td>-295</td>\n",
       "      <td>A</td>\n",
       "      <td>-264</td>\n",
       "      <td>A</td>\n",
       "      <td>-376</td>\n",
       "      <td>A</td>\n",
       "      <td>-419</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-51</td>\n",
       "      <td>A</td>\n",
       "      <td>-139</td>\n",
       "      <td>A</td>\n",
       "      <td>-367</td>\n",
       "      <td>A</td>\n",
       "      <td>-188</td>\n",
       "      <td>A</td>\n",
       "      <td>-407</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7124</th>\n",
       "      <td>PTGER3 Prostaglandin E receptor 3 (subtype EP3...</td>\n",
       "      <td>X83863_at</td>\n",
       "      <td>793</td>\n",
       "      <td>A</td>\n",
       "      <td>782</td>\n",
       "      <td>A</td>\n",
       "      <td>1138</td>\n",
       "      <td>A</td>\n",
       "      <td>627</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>279</td>\n",
       "      <td>A</td>\n",
       "      <td>737</td>\n",
       "      <td>A</td>\n",
       "      <td>588</td>\n",
       "      <td>A</td>\n",
       "      <td>1170</td>\n",
       "      <td>A</td>\n",
       "      <td>2315</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7125</th>\n",
       "      <td>HMG2 High-mobility group (nonhistone chromosom...</td>\n",
       "      <td>Z17240_at</td>\n",
       "      <td>329</td>\n",
       "      <td>A</td>\n",
       "      <td>295</td>\n",
       "      <td>A</td>\n",
       "      <td>777</td>\n",
       "      <td>P</td>\n",
       "      <td>170</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>51</td>\n",
       "      <td>A</td>\n",
       "      <td>227</td>\n",
       "      <td>A</td>\n",
       "      <td>361</td>\n",
       "      <td>A</td>\n",
       "      <td>284</td>\n",
       "      <td>A</td>\n",
       "      <td>250</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7126</th>\n",
       "      <td>RB1 Retinoblastoma 1 (including osteosarcoma)</td>\n",
       "      <td>L49218_f_at</td>\n",
       "      <td>36</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>41</td>\n",
       "      <td>A</td>\n",
       "      <td>-50</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>A</td>\n",
       "      <td>-9</td>\n",
       "      <td>A</td>\n",
       "      <td>-26</td>\n",
       "      <td>A</td>\n",
       "      <td>39</td>\n",
       "      <td>A</td>\n",
       "      <td>-12</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7127</th>\n",
       "      <td>GB DEF = Glycophorin Sta (type A) exons 3 and ...</td>\n",
       "      <td>M71243_f_at</td>\n",
       "      <td>191</td>\n",
       "      <td>A</td>\n",
       "      <td>76</td>\n",
       "      <td>A</td>\n",
       "      <td>228</td>\n",
       "      <td>A</td>\n",
       "      <td>126</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>2484</td>\n",
       "      <td>P</td>\n",
       "      <td>371</td>\n",
       "      <td>A</td>\n",
       "      <td>133</td>\n",
       "      <td>A</td>\n",
       "      <td>298</td>\n",
       "      <td>A</td>\n",
       "      <td>790</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7128</th>\n",
       "      <td>GB DEF = mRNA (clone 1A7)</td>\n",
       "      <td>Z78285_f_at</td>\n",
       "      <td>-37</td>\n",
       "      <td>A</td>\n",
       "      <td>-14</td>\n",
       "      <td>A</td>\n",
       "      <td>-41</td>\n",
       "      <td>A</td>\n",
       "      <td>-91</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>A</td>\n",
       "      <td>-31</td>\n",
       "      <td>A</td>\n",
       "      <td>-32</td>\n",
       "      <td>A</td>\n",
       "      <td>-3</td>\n",
       "      <td>A</td>\n",
       "      <td>-10</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7129 rows × 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Gene Description Gene Accession Number  \\\n",
       "0                   AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at   \n",
       "1                   AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at   \n",
       "2                   AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   \n",
       "3                   AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at   \n",
       "4                   AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at   \n",
       "...                                                 ...                   ...   \n",
       "7124  PTGER3 Prostaglandin E receptor 3 (subtype EP3...             X83863_at   \n",
       "7125  HMG2 High-mobility group (nonhistone chromosom...             Z17240_at   \n",
       "7126      RB1 Retinoblastoma 1 (including osteosarcoma)           L49218_f_at   \n",
       "7127  GB DEF = Glycophorin Sta (type A) exons 3 and ...           M71243_f_at   \n",
       "7128                          GB DEF = mRNA (clone 1A7)           Z78285_f_at   \n",
       "\n",
       "        1 call    2 call.1     3 call.2    4 call.3  ...    29 call.33   30  \\\n",
       "0    -214    A -139      A   -76      A -135      A  ...    15       A -318   \n",
       "1    -153    A  -73      A   -49      A -114      A  ...  -114       A -192   \n",
       "2     -58    A   -1      A  -307      A  265      A  ...     2       A  -95   \n",
       "3      88    A  283      A   309      A   12      A  ...   193       A  312   \n",
       "4    -295    A -264      A  -376      A -419      A  ...   -51       A -139   \n",
       "...   ...  ...  ...    ...   ...    ...  ...    ...  ...   ...     ...  ...   \n",
       "7124  793    A  782      A  1138      A  627      A  ...   279       A  737   \n",
       "7125  329    A  295      A   777      P  170      A  ...    51       A  227   \n",
       "7126   36    A   11      A    41      A  -50      A  ...     6       A   -9   \n",
       "7127  191    A   76      A   228      A  126      A  ...  2484       P  371   \n",
       "7128  -37    A  -14      A   -41      A  -91      A  ...    -2       A  -31   \n",
       "\n",
       "     call.34   31 call.35    32 call.36    33 call.37  \n",
       "0          A  -32       A  -124       A  -135       A  \n",
       "1          A  -49       A   -79       A  -186       A  \n",
       "2          A   49       A   -37       A   -70       A  \n",
       "3          A  230       P   330       A   337       A  \n",
       "4          A -367       A  -188       A  -407       A  \n",
       "...      ...  ...     ...   ...     ...   ...     ...  \n",
       "7124       A  588       A  1170       A  2315       A  \n",
       "7125       A  361       A   284       A   250       A  \n",
       "7126       A  -26       A    39       A   -12       A  \n",
       "7127       A  133       A   298       A   790       P  \n",
       "7128       A  -32       A    -3       A   -10       A  \n",
       "\n",
       "[7129 rows x 78 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "reserved-indiana",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene Description', 'Gene Accession Number', '1', 'call', '2', 'call.1',\n",
       "       '3', 'call.2', '4', 'call.3', '5', 'call.4', '6', 'call.5', '7',\n",
       "       'call.6', '8', 'call.7', '9', 'call.8', '10', 'call.9', '11', 'call.10',\n",
       "       '12', 'call.11', '13', 'call.12', '14', 'call.13', '15', 'call.14',\n",
       "       '16', 'call.15', '17', 'call.16', '18', 'call.17', '19', 'call.18',\n",
       "       '20', 'call.19', '21', 'call.20', '22', 'call.21', '23', 'call.22',\n",
       "       '24', 'call.23', '25', 'call.24', '26', 'call.25', '27', 'call.26',\n",
       "       '34', 'call.27', '35', 'call.28', '36', 'call.29', '37', 'call.30',\n",
       "       '38', 'call.31', '28', 'call.32', '29', 'call.33', '30', 'call.34',\n",
       "       '31', 'call.35', '32', 'call.36', '33', 'call.37'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ranging-success",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>38</th>\n",
       "      <th>28</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "      <td>7129.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>641.367092</td>\n",
       "      <td>690.246318</td>\n",
       "      <td>698.307897</td>\n",
       "      <td>600.985271</td>\n",
       "      <td>679.532894</td>\n",
       "      <td>564.797728</td>\n",
       "      <td>584.437649</td>\n",
       "      <td>571.359097</td>\n",
       "      <td>789.713705</td>\n",
       "      <td>599.483097</td>\n",
       "      <td>...</td>\n",
       "      <td>514.496704</td>\n",
       "      <td>775.143498</td>\n",
       "      <td>689.248141</td>\n",
       "      <td>626.885959</td>\n",
       "      <td>673.279422</td>\n",
       "      <td>556.463179</td>\n",
       "      <td>718.934493</td>\n",
       "      <td>598.648899</td>\n",
       "      <td>676.920887</td>\n",
       "      <td>723.563473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>2264.294361</td>\n",
       "      <td>2468.814372</td>\n",
       "      <td>2485.656277</td>\n",
       "      <td>2340.047428</td>\n",
       "      <td>2375.895416</td>\n",
       "      <td>2494.604090</td>\n",
       "      <td>2412.812263</td>\n",
       "      <td>2378.780450</td>\n",
       "      <td>2580.157021</td>\n",
       "      <td>2421.156219</td>\n",
       "      <td>...</td>\n",
       "      <td>2440.722824</td>\n",
       "      <td>2676.664777</td>\n",
       "      <td>2543.537830</td>\n",
       "      <td>2473.180838</td>\n",
       "      <td>2413.149603</td>\n",
       "      <td>2376.681824</td>\n",
       "      <td>2533.678058</td>\n",
       "      <td>2405.268550</td>\n",
       "      <td>2436.964933</td>\n",
       "      <td>2507.382019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-19826.000000</td>\n",
       "      <td>-17930.000000</td>\n",
       "      <td>-27182.000000</td>\n",
       "      <td>-23396.000000</td>\n",
       "      <td>-10339.000000</td>\n",
       "      <td>-21658.000000</td>\n",
       "      <td>-24024.000000</td>\n",
       "      <td>-27570.000000</td>\n",
       "      <td>-25171.000000</td>\n",
       "      <td>-12500.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-16281.000000</td>\n",
       "      <td>-27398.000000</td>\n",
       "      <td>-23673.000000</td>\n",
       "      <td>-23645.000000</td>\n",
       "      <td>-20376.000000</td>\n",
       "      <td>-9501.000000</td>\n",
       "      <td>-17580.000000</td>\n",
       "      <td>-25491.000000</td>\n",
       "      <td>-28400.000000</td>\n",
       "      <td>-27811.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>-21.000000</td>\n",
       "      <td>-14.000000</td>\n",
       "      <td>-31.000000</td>\n",
       "      <td>-33.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>-26.000000</td>\n",
       "      <td>-33.000000</td>\n",
       "      <td>-58.000000</td>\n",
       "      <td>-14.000000</td>\n",
       "      <td>-15.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-43.000000</td>\n",
       "      <td>-27.000000</td>\n",
       "      <td>-23.000000</td>\n",
       "      <td>-22.000000</td>\n",
       "      <td>-16.000000</td>\n",
       "      <td>-13.000000</td>\n",
       "      <td>-25.000000</td>\n",
       "      <td>-32.000000</td>\n",
       "      <td>-22.000000</td>\n",
       "      <td>-38.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>159.000000</td>\n",
       "      <td>130.000000</td>\n",
       "      <td>177.000000</td>\n",
       "      <td>139.000000</td>\n",
       "      <td>146.000000</td>\n",
       "      <td>106.000000</td>\n",
       "      <td>134.000000</td>\n",
       "      <td>140.000000</td>\n",
       "      <td>166.000000</td>\n",
       "      <td>103.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>108.000000</td>\n",
       "      <td>144.000000</td>\n",
       "      <td>134.000000</td>\n",
       "      <td>133.000000</td>\n",
       "      <td>150.000000</td>\n",
       "      <td>82.000000</td>\n",
       "      <td>128.000000</td>\n",
       "      <td>107.000000</td>\n",
       "      <td>155.000000</td>\n",
       "      <td>170.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>535.000000</td>\n",
       "      <td>488.000000</td>\n",
       "      <td>610.000000</td>\n",
       "      <td>497.000000</td>\n",
       "      <td>471.000000</td>\n",
       "      <td>401.000000</td>\n",
       "      <td>497.000000</td>\n",
       "      <td>527.000000</td>\n",
       "      <td>609.000000</td>\n",
       "      <td>386.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>396.000000</td>\n",
       "      <td>569.000000</td>\n",
       "      <td>505.000000</td>\n",
       "      <td>490.000000</td>\n",
       "      <td>517.000000</td>\n",
       "      <td>309.000000</td>\n",
       "      <td>488.000000</td>\n",
       "      <td>443.000000</td>\n",
       "      <td>549.000000</td>\n",
       "      <td>649.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>31086.000000</td>\n",
       "      <td>29288.000000</td>\n",
       "      <td>28056.000000</td>\n",
       "      <td>31449.000000</td>\n",
       "      <td>29543.000000</td>\n",
       "      <td>38467.000000</td>\n",
       "      <td>41911.000000</td>\n",
       "      <td>40065.000000</td>\n",
       "      <td>23602.000000</td>\n",
       "      <td>28033.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>61228.000000</td>\n",
       "      <td>37164.000000</td>\n",
       "      <td>32204.000000</td>\n",
       "      <td>29169.000000</td>\n",
       "      <td>29833.000000</td>\n",
       "      <td>30354.000000</td>\n",
       "      <td>25055.000000</td>\n",
       "      <td>28350.000000</td>\n",
       "      <td>25093.000000</td>\n",
       "      <td>32946.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  1             2             3             4             5  \\\n",
       "count   7129.000000   7129.000000   7129.000000   7129.000000   7129.000000   \n",
       "mean     641.367092    690.246318    698.307897    600.985271    679.532894   \n",
       "std     2264.294361   2468.814372   2485.656277   2340.047428   2375.895416   \n",
       "min   -19826.000000 -17930.000000 -27182.000000 -23396.000000 -10339.000000   \n",
       "25%      -21.000000    -14.000000    -31.000000    -33.000000      8.000000   \n",
       "50%      159.000000    130.000000    177.000000    139.000000    146.000000   \n",
       "75%      535.000000    488.000000    610.000000    497.000000    471.000000   \n",
       "max    31086.000000  29288.000000  28056.000000  31449.000000  29543.000000   \n",
       "\n",
       "                  6             7             8             9            10  \\\n",
       "count   7129.000000   7129.000000   7129.000000   7129.000000   7129.000000   \n",
       "mean     564.797728    584.437649    571.359097    789.713705    599.483097   \n",
       "std     2494.604090   2412.812263   2378.780450   2580.157021   2421.156219   \n",
       "min   -21658.000000 -24024.000000 -27570.000000 -25171.000000 -12500.000000   \n",
       "25%      -26.000000    -33.000000    -58.000000    -14.000000    -15.000000   \n",
       "50%      106.000000    134.000000    140.000000    166.000000    103.000000   \n",
       "75%      401.000000    497.000000    527.000000    609.000000    386.000000   \n",
       "max    38467.000000  41911.000000  40065.000000  23602.000000  28033.000000   \n",
       "\n",
       "       ...            35            36            37            38  \\\n",
       "count  ...   7129.000000   7129.000000   7129.000000   7129.000000   \n",
       "mean   ...    514.496704    775.143498    689.248141    626.885959   \n",
       "std    ...   2440.722824   2676.664777   2543.537830   2473.180838   \n",
       "min    ... -16281.000000 -27398.000000 -23673.000000 -23645.000000   \n",
       "25%    ...    -43.000000    -27.000000    -23.000000    -22.000000   \n",
       "50%    ...    108.000000    144.000000    134.000000    133.000000   \n",
       "75%    ...    396.000000    569.000000    505.000000    490.000000   \n",
       "max    ...  61228.000000  37164.000000  32204.000000  29169.000000   \n",
       "\n",
       "                 28            29            30            31            32  \\\n",
       "count   7129.000000   7129.000000   7129.000000   7129.000000   7129.000000   \n",
       "mean     673.279422    556.463179    718.934493    598.648899    676.920887   \n",
       "std     2413.149603   2376.681824   2533.678058   2405.268550   2436.964933   \n",
       "min   -20376.000000  -9501.000000 -17580.000000 -25491.000000 -28400.000000   \n",
       "25%      -16.000000    -13.000000    -25.000000    -32.000000    -22.000000   \n",
       "50%      150.000000     82.000000    128.000000    107.000000    155.000000   \n",
       "75%      517.000000    309.000000    488.000000    443.000000    549.000000   \n",
       "max    29833.000000  30354.000000  25055.000000  28350.000000  25093.000000   \n",
       "\n",
       "                 33  \n",
       "count   7129.000000  \n",
       "mean     723.563473  \n",
       "std     2507.382019  \n",
       "min   -27811.000000  \n",
       "25%      -38.000000  \n",
       "50%      170.000000  \n",
       "75%      649.000000  \n",
       "max    32946.000000  \n",
       "\n",
       "[8 rows x 38 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data.describe()#include='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "furnished-nickname",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 7129 entries, 0 to 7128\n",
      "Data columns (total 78 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   Gene Description       7129 non-null   object\n",
      " 1   Gene Accession Number  7129 non-null   object\n",
      " 2   1                      7129 non-null   int64 \n",
      " 3   call                   7129 non-null   object\n",
      " 4   2                      7129 non-null   int64 \n",
      " 5   call.1                 7129 non-null   object\n",
      " 6   3                      7129 non-null   int64 \n",
      " 7   call.2                 7129 non-null   object\n",
      " 8   4                      7129 non-null   int64 \n",
      " 9   call.3                 7129 non-null   object\n",
      " 10  5                      7129 non-null   int64 \n",
      " 11  call.4                 7129 non-null   object\n",
      " 12  6                      7129 non-null   int64 \n",
      " 13  call.5                 7129 non-null   object\n",
      " 14  7                      7129 non-null   int64 \n",
      " 15  call.6                 7129 non-null   object\n",
      " 16  8                      7129 non-null   int64 \n",
      " 17  call.7                 7129 non-null   object\n",
      " 18  9                      7129 non-null   int64 \n",
      " 19  call.8                 7129 non-null   object\n",
      " 20  10                     7129 non-null   int64 \n",
      " 21  call.9                 7129 non-null   object\n",
      " 22  11                     7129 non-null   int64 \n",
      " 23  call.10                7129 non-null   object\n",
      " 24  12                     7129 non-null   int64 \n",
      " 25  call.11                7129 non-null   object\n",
      " 26  13                     7129 non-null   int64 \n",
      " 27  call.12                7129 non-null   object\n",
      " 28  14                     7129 non-null   int64 \n",
      " 29  call.13                7129 non-null   object\n",
      " 30  15                     7129 non-null   int64 \n",
      " 31  call.14                7129 non-null   object\n",
      " 32  16                     7129 non-null   int64 \n",
      " 33  call.15                7129 non-null   object\n",
      " 34  17                     7129 non-null   int64 \n",
      " 35  call.16                7129 non-null   object\n",
      " 36  18                     7129 non-null   int64 \n",
      " 37  call.17                7129 non-null   object\n",
      " 38  19                     7129 non-null   int64 \n",
      " 39  call.18                7129 non-null   object\n",
      " 40  20                     7129 non-null   int64 \n",
      " 41  call.19                7129 non-null   object\n",
      " 42  21                     7129 non-null   int64 \n",
      " 43  call.20                7129 non-null   object\n",
      " 44  22                     7129 non-null   int64 \n",
      " 45  call.21                7129 non-null   object\n",
      " 46  23                     7129 non-null   int64 \n",
      " 47  call.22                7129 non-null   object\n",
      " 48  24                     7129 non-null   int64 \n",
      " 49  call.23                7129 non-null   object\n",
      " 50  25                     7129 non-null   int64 \n",
      " 51  call.24                7129 non-null   object\n",
      " 52  26                     7129 non-null   int64 \n",
      " 53  call.25                7129 non-null   object\n",
      " 54  27                     7129 non-null   int64 \n",
      " 55  call.26                7129 non-null   object\n",
      " 56  34                     7129 non-null   int64 \n",
      " 57  call.27                7129 non-null   object\n",
      " 58  35                     7129 non-null   int64 \n",
      " 59  call.28                7129 non-null   object\n",
      " 60  36                     7129 non-null   int64 \n",
      " 61  call.29                7129 non-null   object\n",
      " 62  37                     7129 non-null   int64 \n",
      " 63  call.30                7129 non-null   object\n",
      " 64  38                     7129 non-null   int64 \n",
      " 65  call.31                7129 non-null   object\n",
      " 66  28                     7129 non-null   int64 \n",
      " 67  call.32                7129 non-null   object\n",
      " 68  29                     7129 non-null   int64 \n",
      " 69  call.33                7129 non-null   object\n",
      " 70  30                     7129 non-null   int64 \n",
      " 71  call.34                7129 non-null   object\n",
      " 72  31                     7129 non-null   int64 \n",
      " 73  call.35                7129 non-null   object\n",
      " 74  32                     7129 non-null   int64 \n",
      " 75  call.36                7129 non-null   object\n",
      " 76  33                     7129 non-null   int64 \n",
      " 77  call.37                7129 non-null   object\n",
      "dtypes: int64(38), object(40)\n",
      "memory usage: 4.2+ MB\n"
     ]
    }
   ],
   "source": [
    "train_data.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "vietnamese-proposition",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A', 'P', 'M'], dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data['call.2'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "loving-jesus",
   "metadata": {},
   "source": [
    "The columns named 'call' have 3 possible values: Present, Absent, and Marginal.\n",
    "In [this paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1409797/) the authors explain that the DNA microarray manufacturer Affymetrix uses a non-parametric statistical test (Wilcoxon signed rank test) of whether significantly more perfect matches show more hybridization signal than their corresponding mismatches to produce the detection call (Absent (A), Present (P) or Marginal (M)). Their results show that the use of a threshold fraction of Present detection calls (derived by MAS5) provides a simple method that effectively eliminates from analysis probe sets that are unlikely to be reliable while preserving the most significant probe sets and those turned on or off; it thereby increases the ratio of true positives to false positives.\n",
    "\n",
    "For these reasons, I will work only with those rows where `'call' == 'P'`\n",
    "\n",
    "[More information about this here.](https://www.kaggle.com/crawford/gene-expression/discussion/120087)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "coral-premises",
   "metadata": {},
   "source": [
    "what is each row? a gene\n",
    "\n",
    "what is each column? for the ones named with numbers: each column is a patient\n",
    "                     for the ones named \"call.#\": how the gene was classified as A, P or M >>> what are these?\n",
    "                     `train_data['call.2'].unique()`<br>\n",
    "                     `array(['A', 'P', 'M'], dtype=object)`<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "welcome-tennis",
   "metadata": {},
   "source": [
    "I want to keep only 'P' columns\n",
    "\n",
    "\n",
    "should I reorder the columns?\n",
    "\n",
    "do I have duplicates?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "mineral-knowing",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "civilian-response",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "responsible-tampa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "renewable-captain",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "talented-report",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "confident-buying",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene Description', 'Gene Accession Number', '39', 'call', '40',\n",
       "       'call.1', '42', 'call.2', '47', 'call.3', '48', 'call.4', '49',\n",
       "       'call.5', '41', 'call.6', '43', 'call.7', '44', 'call.8', '45',\n",
       "       'call.9', '46', 'call.10', '70', 'call.11', '71', 'call.12', '72',\n",
       "       'call.13', '68', 'call.14', '69', 'call.15', '67', 'call.16', '55',\n",
       "       'call.17', '56', 'call.18', '59', 'call.19', '52', 'call.20', '53',\n",
       "       'call.21', '51', 'call.22', '50', 'call.23', '54', 'call.24', '57',\n",
       "       'call.25', '58', 'call.26', '60', 'call.27', '61', 'call.28', '65',\n",
       "       'call.29', '66', 'call.30', '63', 'call.31', '64', 'call.32', '62',\n",
       "       'call.33'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "built-sussex",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Description</th>\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>39</th>\n",
       "      <th>call</th>\n",
       "      <th>40</th>\n",
       "      <th>call.1</th>\n",
       "      <th>42</th>\n",
       "      <th>call.2</th>\n",
       "      <th>47</th>\n",
       "      <th>call.3</th>\n",
       "      <th>...</th>\n",
       "      <th>65</th>\n",
       "      <th>call.29</th>\n",
       "      <th>66</th>\n",
       "      <th>call.30</th>\n",
       "      <th>63</th>\n",
       "      <th>call.31</th>\n",
       "      <th>64</th>\n",
       "      <th>call.32</th>\n",
       "      <th>62</th>\n",
       "      <th>call.33</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFFX-BioB-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-5_at</td>\n",
       "      <td>-342</td>\n",
       "      <td>A</td>\n",
       "      <td>-87</td>\n",
       "      <td>A</td>\n",
       "      <td>22</td>\n",
       "      <td>A</td>\n",
       "      <td>-243</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-62</td>\n",
       "      <td>A</td>\n",
       "      <td>-58</td>\n",
       "      <td>A</td>\n",
       "      <td>-161</td>\n",
       "      <td>A</td>\n",
       "      <td>-48</td>\n",
       "      <td>A</td>\n",
       "      <td>-176</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFFX-BioB-M_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-M_at</td>\n",
       "      <td>-200</td>\n",
       "      <td>A</td>\n",
       "      <td>-248</td>\n",
       "      <td>A</td>\n",
       "      <td>-153</td>\n",
       "      <td>A</td>\n",
       "      <td>-218</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-198</td>\n",
       "      <td>A</td>\n",
       "      <td>-217</td>\n",
       "      <td>A</td>\n",
       "      <td>-215</td>\n",
       "      <td>A</td>\n",
       "      <td>-531</td>\n",
       "      <td>A</td>\n",
       "      <td>-284</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFFX-BioB-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-3_at</td>\n",
       "      <td>41</td>\n",
       "      <td>A</td>\n",
       "      <td>262</td>\n",
       "      <td>A</td>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>-163</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-5</td>\n",
       "      <td>A</td>\n",
       "      <td>63</td>\n",
       "      <td>A</td>\n",
       "      <td>-46</td>\n",
       "      <td>A</td>\n",
       "      <td>-124</td>\n",
       "      <td>A</td>\n",
       "      <td>-81</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFFX-BioC-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-5_at</td>\n",
       "      <td>328</td>\n",
       "      <td>A</td>\n",
       "      <td>295</td>\n",
       "      <td>A</td>\n",
       "      <td>276</td>\n",
       "      <td>A</td>\n",
       "      <td>182</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>141</td>\n",
       "      <td>A</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>146</td>\n",
       "      <td>A</td>\n",
       "      <td>431</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFFX-BioC-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-3_at</td>\n",
       "      <td>-224</td>\n",
       "      <td>A</td>\n",
       "      <td>-226</td>\n",
       "      <td>A</td>\n",
       "      <td>-211</td>\n",
       "      <td>A</td>\n",
       "      <td>-289</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-256</td>\n",
       "      <td>A</td>\n",
       "      <td>-191</td>\n",
       "      <td>A</td>\n",
       "      <td>-172</td>\n",
       "      <td>A</td>\n",
       "      <td>-496</td>\n",
       "      <td>A</td>\n",
       "      <td>-294</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene Description Gene Accession Number   39 call   40  \\\n",
       "0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -342    A  -87   \n",
       "1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -200    A -248   \n",
       "2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   41    A  262   \n",
       "3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at  328    A  295   \n",
       "4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -224    A -226   \n",
       "\n",
       "  call.1   42 call.2   47 call.3  ...   65 call.29   66 call.30   63 call.31  \\\n",
       "0      A   22      A -243      A  ...  -62       A  -58       A -161       A   \n",
       "1      A -153      A -218      A  ... -198       A -217       A -215       A   \n",
       "2      A   17      A -163      A  ...   -5       A   63       A  -46       A   \n",
       "3      A  276      A  182      A  ...  141       A   95       A  146       A   \n",
       "4      A -211      A -289      A  ... -256       A -191       A -172       A   \n",
       "\n",
       "    64 call.32   62 call.33  \n",
       "0  -48       A -176       A  \n",
       "1 -531       A -284       A  \n",
       "2 -124       A  -81       A  \n",
       "3  431       A    9       A  \n",
       "4 -496       A -294       A  \n",
       "\n",
       "[5 rows x 70 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fresh-street",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
